File: GSE_data_analysis/analyze_GSE75214.py
# -*- coding: utf-8 -*-
"""
GSE75214 基因表达数据分析
- 差异基因分析
- 火山图、热图
- GO/KEGG 富集分析
"""

import os
import sys
import gzip

os.makedirs("data", exist_ok=True)
os.makedirs("results", exist_ok=True)
os.makedirs("figures", exist_ok=True)

import pandas as pd
import numpy as np
from scipy import stats
from statsmodels.stats.multitest import multipletests
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt
import seaborn as sns

plt.rcParams['font.sans-serif'] = ['SimHei', 'Microsoft YaHei', 'DejaVu Sans']
plt.rcParams['axes.unicode_minus'] = False


def download_series_matrix():
    """下载 series matrix 并解析"""
    import urllib.request
    url = "https://ftp.ncbi.nlm.nih.gov/geo/series/GSE75nnn/GSE75214/matrix/GSE75214_series_matrix.txt.gz"
    path = "data/GSE75214_series_matrix.txt.gz"
    if not os.path.exists(path):
        print("  下载 series matrix...")
        urllib.request.urlretrieve(url, path)
    return path


def parse_series_matrix(path):
    """解析 series matrix 文件"""
    meta_lines = []
    table_lines = []
    in_table = False
    with gzip.open(path, 'rt', encoding='utf-8', errors='ignore') as f:
        for line in f:
            line = line.rstrip('\n\r')
            if line.startswith('!'):
                if not in_table:
                    meta_lines.append(line)
                if 'series_matrix_table_begin' in line:
                    in_table = True
            elif 'series_matrix_table_end' in line:
                break
            elif in_table:
                table_lines.append(line)

    def strip_quotes(s):
        return str(s).strip().strip('"')

    # 解析 metadata：Sample_geo_accession 和 Sample_title
    samples = []
    titles = []
    for line in meta_lines:
        if line.startswith('!Sample_geo_accession'):
            parts = line.split('\t')[1:]
            samples = [strip_quotes(p) for p in parts]
        elif line.startswith('!Sample_title'):
            parts = line.split('\t')[1:]
            titles = [strip_quotes(p) for p in parts]

    sample_meta = {}
    for i, s in enumerate(samples):
        sample_meta[s] = {'group': 'Unknown'}
        text = (titles[i] if i < len(titles) else '') + ' '
        # 从 title 推断分组，如 CD_colon_active_1, UC_colon_inactive_1, Control_1
        if 'control' in text.lower() or 'non-ibd' in text.lower() or 'healthy' in text.lower():
            sample_meta[s]['group'] = 'Control'
        elif 'cd_colon_active' in text.lower() or ('crohn' in text.lower() and 'active' in text.lower()):
            sample_meta[s]['group'] = 'CD_active'
        elif 'cd_' in text.lower() or 'crohn' in text.lower():
            sample_meta[s]['group'] = 'CD'
        elif 'uc_colon_active' in text.lower() or ('uc' in text.lower() and 'active' in text.lower()):
            sample_meta[s]['group'] = 'UC_active'
        elif 'uc_' in text.lower() or 'ulcerative' in text.lower():
            sample_meta[s]['group'] = 'UC_inactive'
        elif sample_meta[s]['group'] == 'Unknown':
            sample_meta[s]['group'] = 'Other'

    # 解析表达矩阵
    if not table_lines:
        raise ValueError("No expression table found")
    header = [strip_quotes(x) for x in table_lines[0].split('\t')]
    rows = []
    for line in table_lines[1:]:
        parts = line.split('\t')
        if len(parts) >= len(header):
            rows.append([strip_quotes(p) for p in parts[:len(header)]])
    df = pd.DataFrame(rows, columns=header)
    df = df.set_index(header[0])
    for c in df.columns:
        df[c] = pd.to_numeric(df[c], errors='coerce')
    df = df.dropna(how='all')
    return df, sample_meta


def load_gpl_annotation(gpl_id="GPL6244"):
    """加载平台注释（使用 GEOparse）"""
    try:
        import GEOparse
        gpl = GEOparse.get_GEO(geo=gpl_id, destdir="data")
        if "gene_assignment" in gpl.table.columns:
            def extract_symbol(s):
                if pd.isna(s) or s == "---":
                    return ""
                parts = str(s).split(" /// ")[0].split(" // ")
                return parts[1].strip() if len(parts) >= 2 else ""
            annot = gpl.table[["ID"]].copy()
            annot["gene_symbol"] = gpl.table["gene_assignment"].apply(extract_symbol)
        elif "Gene Symbol" in gpl.table.columns:
            annot = gpl.table[["ID", "Gene Symbol"]].rename(columns={"Gene Symbol": "gene_symbol", "ID": "probe"})
        else:
            return pd.DataFrame()
        annot = annot.rename(columns={"ID": "probe"})
        annot = annot[(annot["gene_symbol"].notna()) & (annot["gene_symbol"] != "") & (annot["gene_symbol"] != "---")]
        return annot
    except Exception as e:
        print(f"  GPL 注释加载失败: {e}")
        return pd.DataFrame()


def main():
    print("=" * 60)
    print("GSE75214 分析开始")
    print("=" * 60)

    # 1. 下载并解析
    print("\n[1/7] 下载 GSE75214 数据...")
    path = download_series_matrix()
    expr, sample_meta = parse_series_matrix(path)
    print(f"  表达矩阵: {expr.shape[0]} 探针 x {expr.shape[1]} 样本")

    meta = pd.DataFrame.from_dict(sample_meta, orient='index')
    meta = meta[meta.index.isin(expr.columns)]
    print(meta['group'].value_counts())

    # 选择分组
    case_label, ctrl_label = "UC_active", "Control"
    case_samples = meta.index[meta["group"] == case_label].intersection(expr.columns)
    ctrl_samples = meta.index[meta["group"] == ctrl_label].intersection(expr.columns)

    if len(case_samples) < 2 or len(ctrl_samples) < 2:
        case_label, ctrl_label = "CD_active", "Control"
        case_samples = meta.index[meta["group"] == case_label].intersection(expr.columns)
        ctrl_samples = meta.index[meta["group"] == ctrl_label].intersection(expr.columns)

    if len(case_samples) < 2 or len(ctrl_samples) < 2:
        case_label, ctrl_label = "CD", "Control"
        case_samples = meta.index[meta["group"] == case_label].intersection(expr.columns)
        ctrl_samples = meta.index[meta["group"] == ctrl_label].intersection(expr.columns)

    print(f"  {case_label}: {len(case_samples)} 样本")
    print(f"  {ctrl_label}: {len(ctrl_samples)} 样本")

    expr = expr[case_samples.tolist() + ctrl_samples.tolist()]

    # 2. 差异分析
    print("\n[3/7] 差异表达分析...")
    if expr.min().min() > 0 and expr.max().max() > 100:
        expr = np.log2(expr + 1)
        print("  已进行 log2 转换")

    results = []
    for probe in expr.index:
        case_vals = expr.loc[probe, case_samples]
        ctrl_vals = expr.loc[probe, ctrl_samples]
        t_stat, p_val = stats.ttest_ind(case_vals, ctrl_vals, equal_var=False, nan_policy="omit")
        log2fc = case_vals.mean() - ctrl_vals.mean()
        results.append({
            "probe": probe,
            "log2FC": log2fc,
            "pvalue": p_val,
            "mean_case": case_vals.mean(),
            "mean_ctrl": ctrl_vals.mean(),
        })

    deg = pd.DataFrame(results).dropna(subset=["pvalue"])
    deg["padj"] = multipletests(deg["pvalue"], method="fdr_bh")[1]

    padj_cut, lfc_cut = 0.05, 1.0
    deg["sig"] = (deg["padj"] < padj_cut) & (deg["log2FC"].abs() >= lfc_cut)
    print(f"  显著差异探针数: {deg['sig'].sum()}")

    # 3. 探针注释
    print("\n[4/7] 探针注释...")
    annot = load_gpl_annotation()
    if annot.empty:
        annot = pd.DataFrame({'probe': expr.index, 'gene_symbol': expr.index})
    else:
        annot['gene_symbol'] = annot['gene_symbol'].astype(str).str.split(' /// ').str[0].str.strip()
        annot = annot[(annot['gene_symbol'].notna()) & (annot['gene_symbol'] != '') & (annot['gene_symbol'] != 'nan')]

    deg["probe"] = deg["probe"].astype(str)
    annot["probe"] = annot["probe"].astype(str)
    deg_annot = deg.merge(annot, on="probe", how="left")
    deg_annot = deg_annot[(deg_annot["gene_symbol"].notna()) & (deg_annot["gene_symbol"] != "")]

    sig_deg = deg_annot[deg_annot["sig"]].copy()
    up_genes = sig_deg[sig_deg["log2FC"] > 0]["gene_symbol"].drop_duplicates().tolist()
    down_genes = sig_deg[sig_deg["log2FC"] < 0]["gene_symbol"].drop_duplicates().tolist()
    print(f"  上调基因: {len(up_genes)}, 下调基因: {len(down_genes)}")

    deg_annot.to_csv("results/GSE75214_DEG_gene_level.csv", index=False)

    # 4. 火山图
    print("\n[5/7] 绘制火山图...")
    volcano = deg_annot.copy()
    volcano["-log10_padj"] = -np.log10(np.maximum(volcano["padj"].values, 1e-300))

    fig, ax = plt.subplots(figsize=(8, 6))
    is_sig = volcano["sig"]
    ax.scatter(volcano.loc[~is_sig, "log2FC"], volcano.loc[~is_sig, "-log10_padj"],
               s=8, c="grey", alpha=0.4, label="Not significant")
    ax.scatter(volcano.loc[is_sig, "log2FC"], volcano.loc[is_sig, "-log10_padj"],
               s=8, c="red", alpha=0.7, label="Significant")
    ax.axhline(-np.log10(padj_cut), color="blue", linestyle="--", alpha=0.6)
    ax.axvline(lfc_cut, color="black", linestyle="--", alpha=0.6)
    ax.axvline(-lfc_cut, color="black", linestyle="--", alpha=0.6)
    ax.set_xlabel("log2FC")
    ax.set_ylabel("-log10(padj)")
    ax.set_title(f"GSE75214 Volcano: {case_label} vs {ctrl_label}")
    ax.legend()
    plt.tight_layout()
    plt.savefig("figures/GSE75214_volcano.png", dpi=300)
    plt.close()
    print("  已保存 figures/GSE75214_volcano.png")

    # 5. 热图
    print("\n[6/7] 绘制热图...")
    topN = min(50, len(sig_deg))
    if topN > 0:
        top_genes = sig_deg.sort_values("padj").drop_duplicates("gene_symbol").head(topN)["gene_symbol"].tolist()
        probe_to_gene = annot.set_index("probe")["gene_symbol"].to_dict() if not annot.empty else {}
        expr_gene = expr.copy()
        expr_gene["gene_symbol"] = expr_gene.index.map(lambda x: probe_to_gene.get(x, x))
        expr_gene = expr_gene.dropna(subset=["gene_symbol"])
        expr_gene = expr_gene.groupby("gene_symbol").mean()

        available = [g for g in top_genes if g in expr_gene.index]
        if available:
            heatmat = expr_gene.loc[available, case_samples.tolist() + ctrl_samples.tolist()]
            std = heatmat.std(axis=1).replace(0, 1e-8)
            heatmat_z = (heatmat - heatmat.mean(axis=1).values.reshape(-1, 1)) / std.values.reshape(-1, 1)
            col_colors = pd.Series(index=heatmat_z.columns, dtype=str)
            col_colors[case_samples] = "#e74c3c"
            col_colors[ctrl_samples] = "#3498db"

            g = sns.clustermap(heatmat_z, cmap="RdBu_r", center=0, col_colors=col_colors,
                               xticklabels=False, yticklabels=True, figsize=(10, 10))
            g.fig.suptitle(f"GSE75214 Top{len(available)} DEG Heatmap", y=1.02)
            plt.savefig("figures/GSE75214_heatmap.png", dpi=300, bbox_inches="tight")
            plt.close()
            print("  已保存 figures/GSE75214_heatmap.png")

    # 6. GO/KEGG 富集
    print("\n[7/7] GO/KEGG 富集分析...")
    try:
        import gget
        max_genes = 200
        all_sig_genes = up_genes + down_genes
        genes_for_enrich = list(dict.fromkeys(all_sig_genes))[:max_genes]

        if len(genes_for_enrich) >= 5:
            go_res = gget.enrichr(genes_for_enrich, database="ontology")
            kegg_res = gget.enrichr(genes_for_enrich, database="pathway")

            if go_res is not None and len(go_res) > 0:
                go_res.to_csv("results/GSE75214_GO_enrichment.csv", index=False)
                print(f"  GO 富集: {len(go_res)} 条结果 -> results/GSE75214_GO_enrichment.csv")
            if kegg_res is not None and len(kegg_res) > 0:
                kegg_res.to_csv("results/GSE75214_KEGG_enrichment.csv", index=False)
                print(f"  KEGG 富集: {len(kegg_res)} 条结果 -> results/GSE75214_KEGG_enrichment.csv")
        else:
            print("  显著基因数不足，跳过富集分析")
    except Exception as e:
        print(f"  富集分析跳过: {e}")

    print("\n" + "=" * 60)
    print("分析完成")
    print("=" * 60)
    return deg_annot, sig_deg


if __name__ == "__main__":
    main()

File: GSE_data_analysis/analyze_GSE75214.R
# -*- coding: utf-8 -*-
#
# GSE75214 基因表达数据分析（R 版本）
# 1) 读取/下载 GEO 数据（series matrix）
# 2) 根据样本 title 推断分组（Control / UC_active / CD_active / CD ...）
# 3) 每个探针做 Welch t-test 得到 log2FC、pvalue，并做 BH 校正
# 4) 探针到 gene symbol 注释（GPL6244）
# 5) 火山图、热图
# 6) GO/KEGG 富集分析（clusterProfiler）
#
# 依赖包：
#   GEOquery, ggplot2, dplyr, pheatmap, clusterProfiler, org.Hs.eg.db
#
# 运行示例：
#   Rscript analyze_GSE75214.R

rm(list = ls())

safe_require <- function(pkg) {
  if (!requireNamespace(pkg, quietly = TRUE)) {
    install.packages(pkg, repos = "https://cloud.r-project.org")
  }
}

safe_require("GEOquery")
safe_require("ggplot2")
safe_require("dplyr")
safe_require("pheatmap")
safe_require("clusterProfiler")
safe_require("org.Hs.eg.db")

library(GEOquery)
library(ggplot2)
library(dplyr)
library(pheatmap)
library(clusterProfiler)
library(org.Hs.eg.db)

# -------- 基础路径 --------
script_dir <- tryCatch({
  normalizePath(dirname(sys.frame(1)$ofile))
}, error = function(e) {
  getwd()
})
setwd(script_dir)

dir.create("data", showWarnings = FALSE, recursive = TRUE)
dir.create("results", showWarnings = FALSE, recursive = TRUE)
dir.create("figures", showWarnings = FALSE, recursive = TRUE)

cat(strrep("=", 60), "\n", sep = "")
cat("GSE75214 分析开始\n")
cat(strrep("=", 60), "\n", sep = "")

infer_group <- function(title) {
  t <- tolower(ifelse(is.na(title), "", title))
  if (grepl("control", t) || grepl("non-ibd", t) || grepl("healthy", t)) {
    return("Control")
  }
  if (grepl("cd_colon_active", t) || (grepl("crohn", t) && grepl("active", t))) {
    return("CD_active")
  }
  if (grepl("cd_", t) || grepl("crohn", t)) {
    return("CD")
  }
  if (grepl("uc_colon_active", t) || (grepl("uc", t) && grepl("active", t))) {
    return("UC_active")
  }
  if (grepl("uc_", t) || grepl("ulcerative", t)) {
    return("UC_inactive")
  }
  return("Other")
}

extract_titles_from_pdata <- function(expr_set) {
  pdata <- pData(expr_set)
  # 常见列：title、source_name_ch1、characteristics_ch1...
  if ("title" %in% colnames(pdata)) return(as.character(pdata$title))
  if ("source_name_ch1" %in% colnames(pdata)) return(as.character(pdata$source_name_ch1))
  # fallback：用样本名
  return(rownames(pdata))
}

load_expression_matrix <- function(gse_id = "GSE75214") {
  # GEOquery 会自动下载并解析 series matrix
  gse <- getGEO(gse_id, GSEMatrix = TRUE, getGPL = TRUE, AnnotGPL = FALSE)
  # GSE 可能是列表（多个平台），这里取第一个 ExpressionSet
  if (is.list(gse)) expr_set <- gse[[1]] else expr_set <- gse
  expr <- exprs(expr_set)
  expr <- as.matrix(expr)
  return(list(expr = expr, expr_set = expr_set))
}

load_gpl_annotation <- function(gpl_id = "GPL6244") {
  gpl <- getGEO(gpl_id, AnnotGPL = FALSE, getGPL = TRUE)
  gpl_tab <- Table(gpl)

  # python 里优先 gene_assignment，再解析取第一个/第二段里的 gene symbol
  # GEOquery 表字段名可能随版本不同，这里做一些兼容：
  # 典型列：ID、gene_assignment 或 Gene Symbol
  colnames_lower <- tolower(colnames(gpl_tab))

  if ("gene_assignment" %in% colnames_lower) {
    # 映射列名：找到实际列名
    ga_col <- colnames(gpl_tab)[which(colnames_lower == "gene_assignment")][1]
    id_col <- ifelse("id" %in% colnames_lower, colnames(gpl_tab)[which(colnames_lower == "id")][1], "ID")

    extract_symbol <- function(s) {
      if (is.na(s) || s == "---") return("")
      s <- as.character(s)
      # 复刻 python：parts = str(s).split(" /// ")[0].split(" // ")
      parts <- strsplit(s, " /// ", fixed = TRUE)[[1]][1]
      parts2 <- strsplit(parts, " // ", fixed = TRUE)[[1]]
      if (length(parts2) >= 2) return(trimws(parts2[2])) else return("")
    }

    annot <- data.frame(
      probe = as.character(gpl_tab[[id_col]]),
      gene_symbol = vapply(gpl_tab[[ga_col]], extract_symbol, character(1)),
      stringsAsFactors = FALSE
    )
    annot <- annot %>%
      filter(!is.na(gene_symbol), gene_symbol != "", gene_symbol != "nan")
    return(annot)
  }

  if ("gene symbol" %in% colnames(gpl_tab)) {
    annot <- data.frame(
      probe = as.character(gpl_tab$ID),
      gene_symbol = as.character(gpl_tab$`Gene Symbol`),
      stringsAsFactors = FALSE
    )
    annot <- annot %>%
      filter(!is.na(gene_symbol), gene_symbol != "", gene_symbol != "---")
    colnames(annot)[1] <- "probe"
    return(annot)
  }

  # fallback：空注释
  return(data.frame(probe = rownames(gpl_tab), gene_symbol = NA_character_, stringsAsFactors = FALSE)[0, ])
}

dl <- load_expression_matrix("GSE75214")
expr <- dl$expr
expr_set <- dl$expr_set

cat("[1/7] 表达矩阵维度：", nrow(expr), "探针 x ", ncol(expr), "样本\n", sep = "")

titles <- extract_titles_from_pdata(expr_set)
sample_names <- colnames(expr)
groups <- sapply(titles, infer_group)

meta <- data.frame(
  sample = sample_names,
  title = titles,
  group = groups,
  stringsAsFactors = FALSE
)
meta <- meta %>% filter(sample %in% colnames(expr))

cat(meta$group %>% table() %>% as.character(), "\n", sep = "")

case_label <- "UC_active"
ctrl_label <- "Control"
case_samples <- meta %>% filter(group == case_label) %>% pull(sample) %>% intersect(colnames(expr))
ctrl_samples <- meta %>% filter(group == ctrl_label) %>% pull(sample) %>% intersect(colnames(expr))

if (length(case_samples) < 2 || length(ctrl_samples) < 2) {
  case_label <- "CD_active"
  case_samples <- meta %>% filter(group == case_label) %>% pull(sample) %>% intersect(colnames(expr))
}
if (length(case_samples) < 2 || length(ctrl_samples) < 2) {
  case_label <- "CD"
  case_samples <- meta %>% filter(group == case_label) %>% pull(sample) %>% intersect(colnames(expr))
}

if (length(case_samples) < 2 || length(ctrl_samples) < 2) {
  stop("无法找到足够的 case/ctrl 样本（至少各 2 个）。")
}

cat("选择分组：", case_label, " vs ", ctrl_label, "\n", sep = "")
cat("case 样本数：", length(case_samples), "\n", sep = "")
cat("ctrl 样本数：", length(ctrl_samples), "\n", sep = "")

expr <- expr[, c(case_samples, ctrl_samples), drop = FALSE]
case_samples <- c(case_samples)
ctrl_samples <- c(ctrl_samples)

cat("[2/7] 差异表达分析：Welch t-test...\n")
expr_num <- apply(expr, 2, as.numeric)
rownames(expr_num) <- rownames(expr)
colnames(expr_num) <- colnames(expr)

# python 判断：min>0 && max>100 则 log2(expr+1)
if (min(expr_num, na.rm = TRUE) > 0 && max(expr_num, na.rm = TRUE) > 100) {
  expr_num <- log2(expr_num + 1)
  cat("已进行 log2 转换\n")
}

probe_ids <- rownames(expr_num)
results <- vector("list", length(probe_ids))

for (i in seq_along(probe_ids)) {
  probe <- probe_ids[i]
  case_vals <- expr_num[probe, case_samples, drop = TRUE]
  ctrl_vals <- expr_num[probe, ctrl_samples, drop = TRUE]

  # Welch t-test
  tt <- tryCatch(t.test(case_vals, ctrl_vals, var.equal = FALSE), error = function(e) NULL)
  if (is.null(tt)) {
    results[[i]] <- data.frame(
      probe = probe,
      log2FC = NA_real_,
      pvalue = NA_real_,
      mean_case = NA_real_,
      mean_ctrl = NA_real_
    )
  } else {
    results[[i]] <- data.frame(
      probe = probe,
      log2FC = mean(case_vals, na.rm = TRUE) - mean(ctrl_vals, na.rm = TRUE),
      pvalue = tt$p.value,
      mean_case = mean(case_vals, na.rm = TRUE),
      mean_ctrl = mean(ctrl_vals, na.rm = TRUE)
    )
  }
}

deg <- bind_rows(results) %>% filter(!is.na(pvalue))
deg$padj <- p.adjust(deg$pvalue, method = "BH")

padj_cut <- 0.05
lfc_cut <- 1.0
deg$sig <- (deg$padj < padj_cut) & (abs(deg$log2FC) >= lfc_cut)
cat("显著差异探针数：", sum(deg$sig), "\n", sep = "")

cat("[3/7] 探针注释：GPL6244...\n")
annot <- load_gpl_annotation("GPL6244")

if (nrow(annot) == 0) {
  annot <- data.frame(probe = probe_ids, gene_symbol = probe_ids, stringsAsFactors = FALSE)
}

deg_annot <- deg %>%
  mutate(probe = as.character(probe)) %>%
  left_join(annot %>% mutate(probe = as.character(probe)), by = "probe") %>%
  filter(!is.na(gene_symbol), gene_symbol != "")

sig_deg <- deg_annot %>% filter(sig)
up_genes <- sig_deg %>% filter(log2FC > 0) %>% pull(gene_symbol) %>% unique() %>% as.character()
down_genes <- sig_deg %>% filter(log2FC < 0) %>% pull(gene_symbol) %>% unique() %>% as.character()

cat("上调基因：", length(up_genes), " 下调基因：", length(down_genes), "\n", sep = "")

write.csv(deg_annot, file = file.path("results", "GSE75214_DEG_gene_level.csv"), row.names = FALSE)

cat("[4/7] 火山图...\n")
volcano <- deg_annot %>%
  mutate(
    neglog10_padj = -log10(pmax(padj, 1e-300)),
    sig_label = ifelse(sig, "Significant", "Not significant")
  )

vol_plot <- ggplot(volcano, aes(x = log2FC, y = neglog10_padj, color = sig_label)) +
  geom_point(size = 1.3, alpha = 0.8) +
  scale_color_manual(values = c("Not significant" = "grey70", "Significant" = "red")) +
  geom_hline(yintercept = -log10(padj_cut), linetype = "dashed", color = "blue", alpha = 0.6) +
  geom_vline(xintercept = c(-lfc_cut, lfc_cut), linetype = "dashed", color = "black", alpha = 0.6) +
  labs(x = "log2FC", y = "-log10(padj)", title = paste0("GSE75214 Volcano: ", case_label, " vs ", ctrl_label)) +
  theme_bw() +
  theme(plot.title = element_text(hjust = 0.5))

ggsave(filename = file.path("figures", "GSE75214_volcano.png"), plot = vol_plot, dpi = 300, width = 8, height = 6)
cat("已保存 figures/GSE75214_volcano.png\n")

cat("[5/7] 热图...\n")
topN <- min(50, length(unique(sig_deg$gene_symbol)))

if (topN > 0) {
  top_genes <- sig_deg %>%
    arrange(padj) %>%
    distinct(gene_symbol, .keep_all = TRUE) %>%
    slice_head(n = topN) %>%
    pull(gene_symbol)

  probe_to_gene <- annot %>% select(probe, gene_symbol)
  probe_gene_map <- probe_to_gene$gene_symbol[match(rownames(expr_num), probe_to_gene$probe)]

    valid <- !is.na(probe_gene_map) & probe_gene_map != ""
    expr_valid <- expr_num[valid, , drop = FALSE]              # probes x samples
    gene_valid <- probe_gene_map[valid]                        # length = n_probes_valid

    # 按 gene_symbol 对探针做均值合并，得到 gene x samples
    expr_valid_df <- data.frame(
      Gene = gene_valid,
      expr_valid,
      check.names = FALSE
    )
    expr_gene_df <- expr_valid_df %>%
      group_by(Gene) %>%
      summarise(across(where(is.numeric), mean))

    expr_gene_mat <- as.matrix(expr_gene_df[, -1, drop = FALSE])
    rownames(expr_gene_mat) <- expr_gene_df$Gene
    colnames(expr_gene_mat) <- colnames(expr_valid)

  available <- intersect(top_genes, rownames(expr_gene_mat))
  if (length(available) > 0) {
    heatmat <- expr_gene_mat[available, c(case_samples, ctrl_samples), drop = FALSE]

    # Z-score：每行标准化
    heatmat_z <- t(scale(t(heatmat), center = TRUE, scale = TRUE))
    heatmat_z[is.na(heatmat_z)] <- 0

    group_vec <- c(rep("Case", length(case_samples)), rep("Ctrl", length(ctrl_samples)))
    names(group_vec) <- c(case_samples, ctrl_samples)

    ann_col <- data.frame(Group = group_vec)
    rownames(ann_col) <- colnames(heatmat_z)
    ann_colors <- list(Group = c("Case" = "#e74c3c", "Ctrl" = "#3498db"))

    png(file.path("figures", "GSE75214_heatmap.png"), width = 1000, height = 1000, res = 300)
    pheatmap(
      heatmat_z,
      color = colorRampPalette(c("blue", "white", "red"))(100),
      cluster_rows = TRUE,
      cluster_cols = TRUE,
      show_colnames = FALSE,
      fontsize_row = 8,
      annotation_col = ann_col,
      annotation_colors = ann_colors,
      border_color = NA
    )
    dev.off()
    cat("已保存 figures/GSE75214_heatmap.png\n")
  } else {
    cat("热图：可用 top genes 为 0，跳过热图\n")
  }
}

cat("[6/7] GO/KEGG 富集分析...\n")
max_genes <- 200
all_sig_genes <- unique(c(up_genes, down_genes))
genes_for_enrich <- head(all_sig_genes, max_genes)

if (length(genes_for_enrich) >= 5) {
  # symbol -> entrez
  gene_map <- bitr(
    genes_for_enrich,
    fromType = "SYMBOL",
    toType = "ENTREZID",
    OrgDb = org.Hs.eg.db
  )

  entrez_ids <- unique(gene_map$ENTREZID)
  if (length(entrez_ids) >= 5) {
    go_res <- enrichGO(
      gene = entrez_ids,
      OrgDb = org.Hs.eg.db,
      keyType = "ENTREZID",
      ont = "BP",
      pAdjustMethod = "BH",
      qvalueCutoff = 0.05,
      readable = TRUE
    )
    if (!is.null(go_res) && nrow(as.data.frame(go_res)) > 0) {
      write.csv(as.data.frame(go_res), file = file.path("results", "GSE75214_GO_enrichment.csv"), row.names = FALSE)
      cat("GO 富集：", nrow(as.data.frame(go_res)), " 条结果 -> results/GSE75214_GO_enrichment.csv\n", sep = "")
    }

    kegg_res <- enrichKEGG(
      gene = entrez_ids,
      organism = "hsa",
      pAdjustMethod = "BH",
      qvalueCutoff = 0.05
    )
    if (!is.null(kegg_res) && nrow(as.data.frame(kegg_res)) > 0) {
      write.csv(as.data.frame(kegg_res), file = file.path("results", "GSE75214_KEGG_enrichment.csv"), row.names = FALSE)
      cat("KEGG 富集：", nrow(as.data.frame(kegg_res)), " 条结果 -> results/GSE75214_KEGG_enrichment.csv\n", sep = "")
    }
  } else {
    cat("富集分析：基因转 Entrez 后数量不足，跳过\n")
  }
} else {
  cat("显著基因数不足（<5），跳过富集分析\n")
}

cat(strrep("=", 60), "\n", sep = "")
cat("分析完成\n")
cat(strrep("=", 60), "\n", sep = "")

